### DISCUSSION TITLE
How to start debugging solver failure with non-descriptive message

### DISCUSSION THREAD

**Discussion by ladmanj** [OpenModelica/OpenModelica#10979] [Q&A]
Hello dear friends,

my simulation fails in a way depending on a particular configuration with either a comprehensible message  like timestep is too small at ..., or ends with something like this:

```
The initialization finished successfully without homotopy method.
DDASSL had repeated error test failures on the last attempted step.
can't continue. time = 0.602000
model terminate | Integrator failed. | Simulation terminated at time 0.602
### STATISTICS ###
DASKR--  AT T (=R1) AND STEPSIZE H (=R2) THE                                           In above,  R1 =   6.0200000016020E-01   R2 =   4.3679379435198E-16 DASKR--  ERROR TEST FAILED REPEATEDLY OR WITH ABS(H)=HMIN                        
Simulation process failed. Exited with code 0xfffffffffffffff9.
```

I absolutely can't decode what part of my design is causing the problem.
I'm also unable to find any information on how to start debugging that.
I'm still very new to openmodelica and have neither any experience with other simulations other than Spice for electronic circuits.

Thank you very much for any help

**Comment by casella** [comment]
@ladmanj I am simpathetic with your frustration. One of our goals is to improve the debugging experience with OpenModelica, which is not trivial because of the distance between declarative, equation-based Modelica code and runtime code (which is the thing that fails). See [paper1](http://dx.doi.org/10.3384/ecp12076443), [paper2](http://dx.doi.org/10.3384/ECP14096195), [library1](https://github.com/casella/FailureModes).

In the next 12 months we got some funding to try to improve the situation by bringing more runtime info into the debugger. Stay tuned.

Regarding your specific issue, at time 0.602 the ODE solver (DASKR in this case), which is a variable step-size integrator with error control, repeatedly tries to reduce the time step so that the estimated integration error is less than the tolerance, but when it gets to a time step of 4e-16 it fails, because it is below machine-epsilon.

I guess there is some discontinuity or singularity at that point. Try simulating until 0.62 and plot all the derivatives, if one goes through the roof you have a trail to follow for debugging.

Good luck!

**Comment by ladmanj** [comment]
Hi Francesco,
Thank you for your response.
I guess the same thing as you do, but I can't find that variable or 
derivative which is heading to hell.

But I will try harder, maybe exporting the traces and analyzing it in 
gnu-octave or something like that.

J.

**Comment by casella** [comment]
You can filter the simulation results with the "der" string and then plot them manually (if they are not that many) or use an octave/python script to do that on the .mat file.

Normally what you see in these cases is a lot of chattering on at least one derivative, i.e. a black vertical band in the plot, made by a lot of switching between two different values.

This should eventually be done automatically by the debugger, but we're not there yet.

**Reply by ladmanj** [reply]
Thank you, I don't see any chattering. There is, however, the last step, where the derivative changes immediately by an overwhelming amount.
Another problem is that the values in the plot sometimes show defects, that seem to me as software bugs.
Please see the picture and notice the negative time and that the time value exponent seems to be the same as the amplitude-related one.

![image](https://github.com/OpenModelica/OpenModelica/assets/28023956/ac31c200-7553-46ba-97cb-9360a0edb42b)

**Reply by ladmanj** [reply]
Hi,
I have proven, that the problem lies somewhere in the MAT file creation or readout. When switched to CSV file format, the process is much slower, but successful in the end.
Thank you for further analysis and for fixing it in the foreseeable future. 
J.
